pkg/storage/unified/resource/resource.proto

```diff
@@ -0,0 +1,484 @@
+  // The resource version
+  int64 resource_version = 1;
+
+  // Full kubernetes json bytes (although the resource version may not be accurate)
+  bytes value = 2;
+
+  // Operation
+  ResourceOperation operation = 3;
+
+  // The resource has an attached blob
+  bool has_blob = 4;
+}
+
+// The history and trash commands need access to commit messages
+message ResourceMeta {
+  // The resource version
+  int64 resource_version = 1;
+
+  // The optional commit message
+  ResourceOperation operation = 2;
+
+  // Size of the full resource body
+  int32 size = 3;
+
+  // Hash for the resource
+  string hash = 4;
+
+  // The optional commit message
+  string message = 5;
+
+  // The kubernetes metadata section (not the full resource)
+  // https://github.com/kubernetes/kubernetes/blob/v1.30.2/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1496
+  bytes partial_object_meta = 6;
+
+  // The resource has an attached blob
+  bool has_blob = 7;
+}
+
+// Basic blob metadata
+message BlobInfo {
+  // System identifier
+  string path = 1;
+
+  // Content Length
+  int64 size = 2;
+
+  // Content type header
+  string content_type = 3;
+
+  // Content hash used for an etag
+  string hash = 4;
+}
+
+// Status structure is copied from:
+// https://github.com/kubernetes/apimachinery/blob/v0.30.1/pkg/apis/meta/v1/generated.proto#L979
+message StatusResult {
+  // Status of the operation.
+  // One of: "Success" or "Failure".
+  // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+  // +optional
+  string status = 1;
+  // A human-readable description of the status of this operation.
+  // +optional
+  string message = 2;
+  // A machine-readable description of why this operation is in the
+  // "Failure" status. If this value is empty there
+  // is no information available. A Reason clarifies an HTTP status
+  // code but does not override it.
+  // +optional
+  string reason = 3;
+  // Suggested HTTP return code for this status, 0 if not set.
+  // +optional
+  int32 code = 4;
+}
+
+message LinkBlob {
+  enum Action {
+    UNKNOWN = 0;
+    // Upload raw bytes
+    UPLOAD = 1;
+    // Keep the existing blob (valid for updates)
+    KEEP = 2;
+    // Do not keep the existing version (same as not sending LinkBlob, only valid for updates)
+    REMOVE = 3;
+    // TODO... support presigned uploads
+  }
+
+  // Content type header
+  string content_type = 1;
+
+  // Raw value to write
+  bytes value = 2;
+}
+
+// ----------------------------------
+// CRUD Objects
+// ----------------------------------
+
+message CreateRequest {
+  // Requires group+resource to be configuired
+  // If name is not set, a unique name will be generated
+  // The resourceVersion should not be set
+  ResourceKey key = 1;
+  
+  // The resource JSON.  
+  bytes value = 2;
+
+  // Optional commit message
+  string message = 3;
+
+  // Optionally include a large binary object
+  LinkBlob blob = 4;
```

Comment:
Rethinking this. Let's make this this is the UX we want.

Currently, blob and resource are tightly coupled. One can only upload a Blob by call the CreateResource. This lead to:
- Update is a bit weird. We have to specify an action (UPDATE/KEEP/ELSE)
- One the k8s side, what will be the UX like ? Will the BlobUrl be part of the GetResponse ? 

Maybe we should fully decouple those two:
- Provide a Get/Update/Delete Blob methods.
- On the storage side, we inlude the information about the blob (has_blob/link) in the object meta (annotation?)

Reply:
+1, it seems sensible to just expose blob as a subresource with standard semantics.

Could details about the blob be returned as part of Status?

Reply:
Yes -- something awkward here -- i'll iterate some more.  I don't think we want to futz with status since the schema depends on each resource, but marking the existence in annotations seems reasonable